### ISSUE TITLE
Asks/Questions

### ISSUE THREAD

**Issue by ekiru** [glitch-soc/mastodon#205]
Tumblr offers a feature called asks, and there are a few other sites (curious cat, sarahah, maybe others) which offer it as the whole idea of the site.

In case folks aren't aware: the feature is that other people can send you little messages containing questions, and then they go to your inbox and you can choose to ignore them, delete them, answer them publicly (in which case anyone can see them and the question and answer go out onto your followers' dashboards/home timelines/etc.), or answer them privately. These asks can be sent either as one's self or anonymously on most of these platforms.

We would need to be pretty careful in designing and implementing this feature if we do decide to do so because, although anonymous asks can be fun to receive from your friends, they can also provide a lot of room for abuse, depending on how they're implemented and what mechanisms we provide to protect against them. (I think that non-anon asks don't add any additional abuse vectors over DMs as long as we allow the asker an option to require only private replies, but if I'm missing anything there, I'd be grateful to have it pointed out.)

If we provide a way for people to send messages anonymously, abusive people will sometimes try to use it to send harmful or unkind messages. This is the main abuse vector that we need to be mindful of no matter how we choose to design the feature. On Tumblr, when a controversial post gets spread around, the author will often get lots of abusive (often anonymous) asks. As well, people who don't like a person for whatever reason will send them hurtful anons sometimes. Tumblr provides only very blunt tools for protecting one's self against these patterns of abuse, like blocking (anon) asks entirely. If we decide to implement this feature, we will need to provide better and finer-grained tools for individual people on Mastodon instances to protect themselves, as well as for instance admins to protect the people on their instances, and we should do so simultaneously with implementing the feature itself.

I brought this up on masto and a few others discussed possible approaches with me, but it was a while ago so I don't have the thread handy. I'll try to track it down sometime later. But I'll discuss some of the ideas for implementing this and protections against abuse I remember being brought up then.

The first approach was to have anonymous asks be submitted on some unauthenticated page on the recipient's instance and to allow recipients and their admins to block anons based on IP addresses/IP ranges/etc. (of course in addition to allowing them to completely refuse to receive anons). This is probably too incomplete a protection given the ease of evading IP blocks and potential for collateral damage due to shared external IP addresses.

The next approach, I think @hoodiek suggested it, was for even anon asks to be sent from your original server with identifying information that would just not be shown. People could then block individual anonymous askers or potentially even their instances or limit anonymous asks to their followers, admins could see if particular askers were causing problems, and we could prevent blocked people from sending even anonymous asks and hide even anonymous asks from muted people. This would also allow private replies to anon asks (and getting a notification for replies to your anon asks in general), which are features that Tumblr at least lacks. I'm not sure if this should expose which asks come from the same asker to recipients automatically. This approach does have the drawback of leaving the anonymity completely at the discretion of the receiving server, but perhaps that's not such a problem if we communicate that clearly.

Another approach, suggested by @MightyPork IIRC, was a possibly decentralized bot that would receive asks and forward them on to the recipient without identifying information, and allow the recipient to tell the bot to block messages from a particular asker. This has the same anonymity problem as the previous idea but it being up to the bot (and the admin of the bot's instance) instead of the recipient's instance admin. As well, if the bot allows sending asks to remote people (from the perspective of the bot), then a malicious asker could send abusive asks through each existing instance of the bot and the recipient would have to tell every single instance of the bot to block messages from the asker. I'm kind of skeptical of this approach at this point.

There's probably more than could be said but this issue is rather lengthy and I'd rather go ahead and post it and see if anyone else wants to offer input, whether in the form of additional ways to abuse any of these approaches, additional abuse protections we could use, other approaches, feedback on which trade-offs are worthwhile, etc.

(btw as far as the implementation details of representing asks/answers in ActivityPub go, there's already a suggestion in the standards in the form of the Question activity type: see https://www.w3.org/TR/activitystreams-vocabulary/#questions which discusses both this sort of thing and polls)

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by subversive-owl** [comment]
a nice setting would be "askable" or such in the options so that people can opt-in if they want to enable the feature.

**Comment by nightpool** [comment]
I believe I suggested the second approach, which is how I plan to implement it on my tumblr-cone-in-progress, when I get there. My current idea is that all anonymous messages would come from a special "anonymous" actor on the sending server-side, so that anonymity is fully maintained per-server (but the sending server would still maintain a record of accounts -> asks to make blocking work)

this could be extended by using rotating anonymous identifiers, like chans have.

**Comment by aschmitz** [comment]
Opting in to (or out from, probably the same default as whatever the settings for non-following direct messages is) asks seems like a useful feature, yes.

The second approach seems vaguely ideal, given that people already have blocklists of instances they don't wish to federate with, and this could probably lean on those lists (if you don't want to see messages from an instance, it's a safe assumption that you don't want people from there asking you questions, either). I'd agree with @nightpool that the sending instance should ideally anonymize the sending user in the case of anonymous asks, but there's a bit of a question regarding whether or not the random ID should be the same for subsequent asks or not (the former provides better anonymity, the latter allows blocking individuals without necessarily knowing who they are, but sacrifices some anonymity if the questions could be put together to do so).

As a single-user instance, the third option would provide me with better lazy-anonymity, but given that blocks would have to be done in some sort of decentralized manner (or with one centralized "ask anonymizer"), that doesn't seem like a great idea. (On the other hand, even with the second option, it wouldn't be all that hard for someone to stand up an "ask anonymizer" instance that proxied anonymous asks. If there were very few of these, it could still be tractable to block individuals from them, assuming they handled abuse well. If they didn't, they'd probably just earn themselves a place on blocklists as with any other abusive instance.)

**Comment by nightpool** [comment]
yeah some sort of light-weight trusted proxying solution would need to be thought about for single user instances. maybe single-user instances can opt-in to proxying their asks through bigger instances, with that bigger instance having a whitelist of people allowed to proxy through them. that's a pretty simple setup but does require a lot of thinking about ahead of time. hmm.
